src/components/chatbox/message.tsx

```diff
@@ -0,0 +1,251 @@
+	markdown,
+	showActions,
+	conversationId,
+	messageId,
+}: {
+	markdown: string
+	showActions: boolean
+	conversationId: string
+	messageId: string
+}) => {
+	const { session } = useAuthentication()
+	const accessToken = session?.accessToken
+	// Determines green/red button
+	const [rating, setRating] = useState<1 | -1 | 0>(0)
+
+	const handleFeedback = async ({
+		rating,
+		text,
+	}: {
+		rating?: -1 | 1
+		text?: string
+	}) => {
+		const response = await fetch('/api/chat/feedback', {
+			method: 'POST',
+			headers: {
+				'Content-Type': 'application/json',
+				Authorization: `Bearer ${accessToken}`,
+			},
+			body: JSON.stringify({
+				messageId,
+				conversationId,
+				rating,
+				text,
+			}),
+		})
+		return response
+	}
+
+	const [isCopied, setIsCopied] = useState(false)
+	useEffect(() => {
+		let id: ReturnType<typeof setTimeout>
+		if (isCopied) {
+			id = setTimeout(() => {
+				setIsCopied(false)
+			}, 4000)
+		}
+		return () => {
+			clearTimeout(id)
+		}
+	}, [isCopied])
+
+	/** copy rich text; not raw markdown */
+	const handleCopy = () => {
+		// CSS Module classname is a reasonable "hook" to grab the element
+		// rendered by ReactMarkdown.
+		const el = document.getElementsByClassName(s.message_markdown)?.[0]
+
+		if (el) {
+			// https://www.nikouusitalo.com/blog/why-isnt-clipboard-write-copying-my-richtext-html/
+			const clipboardItem = new ClipboardItem({
+				// @ts-expect-error - innerText is available
+				'text/plain': new Blob([el.innerText], { type: 'text/plain' }),
+				'text/html': new Blob([el.outerHTML], { type: 'text/html' }),
+			})
+
+			navigator.clipboard.write([clipboardItem])
+			setIsCopied(true)
+		}
+	}
+
+	return (
+		<div className={cn(s.message, s.message_assistant)}>
+			<IconTile className={cn(s.message_icon)}>
+				<IconWand24 style={{ width: 24, height: 24 }} />
+			</IconTile>
+			<div className={cn(s.message_content)}>
+				<ReactMarkdown
+					className={s.message_markdown}
+					components={{
+						// @ts-expect-error - ignore this type incompatibility
+						pre: MdxPre,
+						p: (props) => (
+							<Text /* Body/200/Medium */
+								size={200}
+								weight="medium"
+								{...props}
+							/>
+						),
+					}}
+					remarkPlugins={[remarkGfm]}
+					rehypePlugins={[[rehypePrism, { ignoreMissing: true }]]}
+				>
+					{markdown}
+				</ReactMarkdown>
+
+				<div
+					className={cn(s.message_AssistantMessageFooter, {
+						[s.message_AssistantMessageFooterHidden]: !showActions,
+					})}
+				>
+					<span className={s.message_divider} />
+					<div className={s.message_actionButtons}>
+						<Button
+							size="small"
+							color="secondary"
+							icon={
+								isCopied ? (
+									<IconCheckSquare24 height={12} width={12} />
+								) : (
+									<IconClipboard24 height={12} width={12} />
+								)
+							}
+							aria-label="Copy to clipboard"
+							text={isCopied ? 'Copied' : 'Copy'}
+							onClick={handleCopy}
+						/>
+
+						<Button
```

Comment:
Can you move these out to their own components so we can make sure they're always visually aligned?

Reply:
@dstaley can you clarify the request here?

- What should be moved?
- What should be visually aligned? (The code itself? UI elements?)

Reply:
Sorry, my message was definitely not well worded! These two buttons are almost visually identical, and they should remain so since they're the same action, just inverse of one another. In instances like this, I like to refactor them out into their own component so that I don't run into a situation where I change the prop on one but not the other. That being said this is certainly just a suggestion, and something that isn't a huge deal since they're right next to each other in code.